File: 0046-permutations/0046-permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    let result = [];
    permuteHelper(nums, 0, result);
    return result;
};

function permuteHelper(nums, start, result) {
    if (start === nums.length) {
        result.push([...nums]);
        return;
    }

    for (let i = start; i < nums.length; i++) {
        [nums[start], nums[i]] = [nums[i], nums[start]]; // swap
        permuteHelper(nums, start + 1, result);
        [nums[start], nums[i]] = [nums[i], nums[start]]; // backtrack
    }
}

File: 0046-permutations/0046-permutations.rs
impl Solution {
    pub fn permute(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut nums = nums;
        let mut result = Vec::new();
        Self::permute_helper(&mut nums, 0, &mut result);
        result
    }

    fn permute_helper(nums: &mut Vec<i32>, start: usize, result: &mut Vec<Vec<i32>>) {
        if start == nums.len() {
            result.push(nums.clone());
            return;
        }

        for i in start..nums.len() {
            nums.swap(start, i);
            Self::permute_helper(nums, start + 1, result);
            nums.swap(start, i); // backtrack
        }
    }
}
